File: templates/problems/11_graphs/03_clone_graph/python/solution.py
from __future__ import annotations

from helpers import Node


def solve_clone_graph(node: Node | None) -> Node | None:
    """Deep copy a graph using DFS with a visited map.

    Time: O(V + E), Space: O(V).
    """
    if node is None:
        return None
    visited: dict[Node, Node] = {}

    def dfs(n: Node) -> Node:
        if n in visited:
            return visited[n]
        clone = Node(n.val)
        visited[n] = clone
        for neighbor in n.neighbors:
            clone.neighbors.append(dfs(neighbor))
        return clone

    return dfs(node)

File: templates/problems/11_graphs/03_clone_graph/typescript/solution.ts
import { Node } from "./helpers";

/**
 * Deep copy a graph using DFS with a visited map.
 * Time: O(V + E), Space: O(V).
 */
export function solveCloneGraph(node: Node | null): Node | null {
    if (node === null) return null;
    const visited = new Map<Node, Node>();

    const dfs = (n: Node): Node => {
        if (visited.has(n)) return visited.get(n)!;
        const clone = new Node(n.val);
        visited.set(n, clone);
        for (const neighbor of n.neighbors) {
            clone.neighbors.push(dfs(neighbor));
        }
        return clone;
    };

    return dfs(node);
}
